src/audio/drc/drc.c

```diff
@@ -0,0 +1,606 @@
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	unsigned char *dst;
+	unsigned char *src;
+	uint32_t offset;
+	size_t size;
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "drc_cmd_set_data(), SOF_CTRL_CMD_BINARY");
+
+		/* Check that there is no work-in-progress previous request */
+		if (cd->config_new && cdata->msg_index == 0) {
+			comp_err(dev, "drc_cmd_set_data(), busy with previous request");
+			return -EBUSY;
+		}
+
+		/* Copy new configuration */
+		if (cdata->msg_index == 0) {
+			/* Allocate buffer for copy of the blob. */
+			size = cdata->num_elems + cdata->elems_remaining;
+			comp_info(dev, "drc_cmd_set_data(), allocating %d for configuration blob",
+				  size);
+			if (size > SOF_DRC_MAX_SIZE) {
+				comp_err(dev, "drc_cmd_set_data(), size exceeds %d",
+					 SOF_DRC_MAX_SIZE);
+				return -EINVAL;
+			}
+
+			cd->config_new = rzalloc(SOF_MEM_ZONE_RUNTIME, 0,
+						 SOF_MEM_CAPS_RAM, size);
+			if (!cd->config_new) {
+				comp_err(dev, "drc_cmd_set_data(): buffer allocation failed");
+				return -EINVAL;
+			}
+
+			offset = 0;
+		} else {
+			assert(cd->config_new);
+			size = cd->config_new->size;
+			offset = size - cdata->elems_remaining -
+				cdata->num_elems;
+		}
+
+		comp_info(dev, "drc_cmd_set_data(), chunk size: %u msg_index %u",
+			  cdata->num_elems, cdata->msg_index);
+		dst = (unsigned char *)cd->config_new;
+		src = (unsigned char *)cdata->data->data;
+
+		/* Just copy the configuration. The EQ will be initialized in
+		 * prepare().
+		 */
+		ret = memcpy_s(dst + offset, size - offset, src,
+			       cdata->num_elems);
+		assert(!ret);
+
+		/* we can check data when elems_remaining == 0 */
+		if (cdata->elems_remaining == 0) {
+			/* If component state is READY we can omit old
+			 * configuration immediately. When in playback/capture
+			 * the new configuration presence is checked in copy().
+			 */
+			if (dev->state == COMP_STATE_READY)
+				drc_free_parameters(&cd->config);
+
+			/* If there is no existing configuration the received
+			 * can be set to current immediately. It will be
+			 * applied in prepare() when streaming starts.
+			 */
+			if (!cd->config) {
+				cd->config = cd->config_new;
+				cd->config_new = NULL;
+			}
+		}
+		break;
+	default:
+		comp_err(dev, "drc_cmd_set_data(), invalid command");
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+static int drc_cmd(struct comp_dev *dev, int cmd, void *data,
+		   int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+	int ret = 0;
+
+	comp_info(dev, "drc_cmd()");
+
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		ret = drc_cmd_set_data(dev, cdata);
+		break;
+	case COMP_CMD_GET_DATA:
+		ret = drc_cmd_get_data(dev, cdata, max_data_size);
+		break;
+	default:
+		comp_err(dev, "drc_cmd(), invalid command");
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+static int drc_trigger(struct comp_dev *dev, int cmd)
+{
+	comp_info(dev, "drc_trigger()");
+
+	return comp_set_state(dev, cmd);
```

Comment:
Should we check the processing func to make sure its set so we don't dereference a null pointer?